### PR TITLE
[None][perf] Fuse DeepSeek-V4 Indexer Q projection with CuTe DSL

### PR DESCRIPTION
**Pull request by mingyangHao**
## Paths compared

- **Original:** FP8 quant + original DeepGEMM Q projection +
  `mlaRoPEInplaceKernel` + `fusedCatFp4Kernel` + output scale copies.
- **Forced fused CuTe:** FP8 quant + one CuTe Q projection/RoPE/FP4-concat
  kernel + output scale copies. It uses the swapped small-M kernel for `M<=8`
  and the native CuTe fused kernel for every `M>=16`.

There is no fallback to the original chain at any measured M. NSYS topology
validation found no standalone RoPE or FP4-concat kernel in any forced-fused
candidate graph.

## Results

Times are the average of the forward/reverse capture medians. The percentage
column is the range from the two independent capture orders; positive means
forced-fused CuTe is faster.

| M | Original output-ready (us) | Forced fused (us) | Output-ready gain |
|---:|---:|---:|---:|
| 1 | 29.576 | 21.864 | +23.21% .. +28.92% |
| 2 | 30.040 | 22.096 | +22.67% .. +30.19% |
| 4 | 31.528 | 24.432 | +21.73% .. +23.29% |
| 8 | 33.008 | 28.824 | +12.33% .. +13.02% |
| 16 | 33.376 | 32.528 | +2.11% .. +2.97% |
| 32 | 34.624 | 32.992 | +3.66% .. +5.77% |
| 64 | 35.576 | 32.360 | +8.80% .. +9.28% |
| 128 | 35.745 | 29.944 | +16.05% .. +16.40% |
| 256 | 23.048 | 18.560 | +19.28% .. +19.67% |
| 512 | 29.712 | 22.672 | +23.68% .. +23.71% |
| 1024 | 42.633 | 31.008 | +27.25% .. +27.28% |
| 2048 | 69.920 | 44.968 | +35.68% .. +35.70% |
| 4096 | 124.408 | 75.984 | +38.91% .. +38.94% |
| 8192 | 245.849 | 143.800 | +41.44% .. +41.58% |
| 16384 | 511.074 | 287.009 | +43.54% .. +44.14% |

## PR Checklist

Please review the following before submitting your PR:
- PR description clearly explains what and why. If using CodeRabbit's summary, please make sure it makes sense.
- PR Follows [TRT-LLM CODING GUIDELINES](https://github.com/NVIDIA/TensorRT-LLM/blob/main/CODING_GUIDELINES.md) to the best of your knowledge.
- Test cases are provided for new code paths (see [test instructions](https://github.com/NVIDIA/TensorRT-LLM/tree/main/tests#1-how-does-the-ci-work))
- If PR introduces API changes, an appropriate PR label is added - either `api-compatible` or `api-breaking`. For `api-breaking`, include `BREAKING` in the PR title.
- Any new dependencies have been scanned for license and vulnerabilities
- [CODEOWNERS](https://github.com/NVIDIA/TensorRT-LLM/blob/main/.github/CODEOWNERS) updated if ownership changes
- Documentation updated as needed
- Update [tava architecture diagram](https://github.com/NVIDIA/TensorRT-LLM/blob/main/.github/tava_architecture_diagram.md) if there is a significant design change in PR.
- The reviewers assigned automatically/manually are appropriate for the PR.

- [x] Please check this after reviewing the above items as appropriate for this PR.

## GitHub Bot Help

To see a list of available CI bot commands, please comment `/bot help`.

## Dev Engineer Review

- **FP8 quantization / scale-layout correctness**
  - `fp8_blockscale_quant_packed.cu` introduces `OutputCuteDslSf` to switch UE8M0 scale output format:
    - **Packed-scale (`false`)**: packs four UE8M0 scale bytes into a 32-bit `packed` value (with K-bound masking) and writes `packed_scale_output`.
    - **CuTe/indexer path (`true`)**: bypasses `packed`; replicates each UE8M0 scale byte into a replicated 32-bit pattern and writes into the SM100 CuTe **128×4 swizzled** layout, emitting only lanes where `lane_id % 8 == 0` and zero-filling padded rows.
  - Host launch wiring now instantiates the kernel with `OutputCuteDslSf=false` for the default launcher and `true` for the CuTe launcher, passing `padded_m` as `scale_leading_dim_uint32`.

- **New SM100-only FP8 quant op (CuTe DSL / UE8M0)**
  - `fp8_blockscale_quant_packed.h` adds `launch_fp8_quantize_1x128_cutedsl_bf16_e4m3(...)` with `padded_m` for the swizzled scale layout.
  - `thop/fp8Quantize.cpp` adds `fp8_quantize_1x128_cutedsl_ue8m0(Tensor)` (SM100-family + input validation, including `k % 128 == 0`), returning `(fp8 values, ue8m0 scale tensor)`.
  - `cpp_custom_ops.py` adds the corresponding fake/meta registration for `trtllm::fp8_quantize_1x128_cutedsl_ue8m0`, computing padded scale shapes from `m` (padded to 128) and the `k`-dependent padded scale dimensions.

- **DeepSeek-V4 indexer-Q fused routing**
  - `deepseek_v4.py` refactors Q handling to unify projection→RoPE→quantization in `_project_and_quantize_q`, using the CuTe fused op `torch.ops.trtllm.cute_dsl_fp8_indexer_q_gemm_rope_fp4_blackwell` when the CuTe/DSv4 constraints are satisfied (MXFP4_BLOCKWISE cache, non-Neox, exact head/rope dims, CuTe cutedsl scale attributes present, SM100f, bf16 path, and `qr` bf16).
  - Otherwise it falls back to the previous sequence (project → apply RoPE in-place → quantize).

- **CuTe DSL runner + operator (Blackwell path)**
  - `cute_dsl_custom_ops.py` adds:
    - `CuteDSLIndexerQBlackwellRunner` (SM100-family gated) with autotuning that:
      - Uses **forced `position_ids = 0` during tuning** to keep RoPE cache lookups consistent.
      - Chooses between **small-M swap_ab tactics** and a native tactic set; includes an eager fallback when `tactic == -1`.
    - Custom op `cute_dsl_fp8_indexer_q_gemm_rope_fp4_blackwell(...)` plus matching fake/meta registration.
  - Dispatch selects wrapper entrypoints (`wrapper_indexer_q_swap_ab` vs `wrapper_indexer_q`) and returns packed FP4 bytes (`int8`) and output scale (`int32`).

- **Blackwell persistent GEMM epilogue integration**
  - `dense_blockscaled_gemm_act_fusion.py` and `dense_blockscaled_gemm_persistent.py` add an `indexer_q_fusion` mode:
    - Plumbs required tensors (`indexer_scale_tensor`, `position_ids_tensor`, `cos_sin_cache_tensor`).
    - Updates epilogue flow to apply rotary embedding in the fused path and alters FP4 SFC/scale handling to support indexer-Q packing/addressing.
    - Introduces wrapper entrypoints including `wrapper_indexer_q_swap_ab`.

- **Weight transform / buffers for fused layout**
  - `tensorrt_llm/_torch/modules/linear.py` extends the FP8 resmoothing/layout flow to generate CUTEDSL-specific indexer-Q buffers (`indexer_q_weight_scale_cutedsl`, `indexer_q_alpha_cutedsl`) when the new fused layout is enabled.

- **CI stability / failure correlation**
  - CI was repeatedly retriggered with bot control commands; some runs for earlier commits reported failures where helper jobs sometimes succeeded.
  - Given the breadth of changes (quant kernel, new fused custom op/runner, and persistent GEMM epilogue wiring), if regressions persist, the highest-risk areas to investigate first are **(1)** scale/layout correctness in the CuTe quant/packing path and **(2)** packed-vs-swizzled FP4/scale bit-exactness in the fused indexer-Q route.

### Test files touched
- `tests/integration/test_lists/test-db/l0_dgx_b200.yml`
  - **Added**: `unittest/_torch/attention/sparse/test_cpp_custom_ops.py::test_cute_dsl_fp8_indexer_q_gemm_rope_fp4_matches_unfused`
  - **Placement/Gating**: under the “DSA FP4 indexer (Blackwell-only)” section; `system_gpu_count: gte: 4, lte: 4`, Ubuntu, `*b200*` GPUs, `stage: pre_merge`, `backend: pytorch`, `orchestrator: mpi`.
- `tests/unittest/_torch/attention/sparse/test_cpp_custom_ops.py`
  - **Added**: `test_cute_dsl_fp8_indexer_q_gemm_rope_fp4_matches_unfused(num_tokens)`
    - Parameterized over `num_tokens` (and performs bit-for-bit equality vs an unfused reference chain).
    - Also asserts expected runner/tactic behavior for small vs larger token counts.

### Coverage / verdict
- **Covered in `test-db` integration list?** Yes (single new entry in `l0_dgx_b200.yml`).
- **Verdict:** needs follow-up (CI L0 instability was reported for other triggered commits, and no additional CBTS/manual-QA coverage status was provided here beyond the new test being added to the L0 list).

### CODE REVIEW

**Review by longlee0622 — APPROVED** [review]
Approve for multi-gpu run

**Review by kris1025 — APPROVED** [review]

File: tensorrt_llm/_torch/custom_ops/cute_dsl_custom_ops.py [hunk 1]
```diff
@@ -3636,6 +3636,320 @@ def _fake_single_b(
...
+            return 4
+        if num_tokens <= 8:
+            return 8
+        if num_tokens <= 16:
+            return 16
+        return max(32, last_positive_power_of_2(num_tokens))
+
+    def _prepare_cutedsl_indexer_q_tuning_inputs(
+            inputs: List[torch.Tensor]) -> List[torch.Tensor]:
+        inputs[3] = torch.zeros_like(inputs[3])
```
Comment (zhaoyangwang-nvidia) [inline tensorrt_llm/_torch/custom_ops/cute_dsl_custom_ops.py]: Magic number here, maybe add some comments.

File: tensorrt_llm/_torch/custom_ops/cute_dsl_custom_ops.py [hunk 2]
```diff
@@ -3636,6 +3636,320 @@ def _fake_single_b(
...
+                    1,
+                    "k",
+                    "k",
+                    "n",
+                )
+            ])
+            return tactics
+
+        @staticmethod
+        def _fallback_tactic(m: int) -> Tuple:
```
Comment (zhaoyangwang-nvidia) [inline tensorrt_llm/_torch/custom_ops/cute_dsl_custom_ops.py]: _fallback_tactic unconditionally picks swap_ab for m <= 8 without the n % 128 == 0 and k % 128 == 0 check that get_valid_tactics applies — the two feasibility judgments disagree

**Review by zhaoyangwang-nvidia — APPROVED** [review]
Approve with little comments.

File: tensorrt_llm/_torch/custom_ops/cute_dsl_custom_ops.py (hunk 1 above)
Comment (mingyangHao) [reply tensorrt_llm/_torch/custom_ops/cute_dsl_custom_ops.py]: Done. Replaced the magic index with a named constant and documented why tuning zeroes position_ids.

File: tensorrt_llm/_torch/custom_ops/cute_dsl_custom_ops.py (hunk 2 above)
Comment (mingyangHao) [reply tensorrt_llm/_torch/custom_ops/cute_dsl_custom_ops.py]: Fixed. get_valid_tactics() and _fallback_tactic() now share the same small-M feasibility check.
